### ISSUE TITLE
Oddity with Encoding Raw Strings

### ISSUE THREAD

**Issue by rpless** [finagle/finch#328]
I've been fixing and expanding the [Finch implementation of the Tech Empower Benchmark](https://github.com/TechEmpower/FrameworkBenchmarks/tree/master/frameworks/Scala/finch).
For the [plain text test](http://example2frameworkbenchmarksdocs.readthedocs.org/en/latest/Project-Information/Framework-Tests/#plaintext), I have a router: 
```scala
  val plaintext: Router[String] = Get / "plaintext" /> "Hello, World!"
```
When I run Tech Empower's tests this yields an error saying that the response has 2 more bytes than expected. When I look at the response I see that there are 2 quotes in the body I don't think should there:
```
HTTP/1.1 200 OK
Content-Type: application/json;charset=utf-8
Date: Wed, 17 Jun 2015 18:09:55 GMT
Content-Length: 15
Server: Finch 0.7.0

"Hello, World!"
```
Are these quotes supposed to be in the response and if so how do I get rid of them? Or is this a bug?

**Comment by vkostyukov** [comment]
Hm. Interesting. I believe these quotes happened after #283. We should probable write a tase case and try to fix it.

**Comment by vkostyukov** [comment]
I hope this doesn't mean that we have the same issue with JSON encoding...

**Comment by rpless** [comment]
It doesn't seem to. I've cleaned up the json test too and it passes their test with flying colors.

**Comment by vkostyukov** [comment]
Just double-checked the demo project. The output JSON is not wrapped with quotes:

```
TTP/1.1 200 OK
Content-Type: application/json;charset=utf-8
Content-Length: 221

[{"id":0,"name":"Tom Paine","tickets":[]},{"id":1,"name":"Benjamin Franklin","tickets":[{"id":2,"label":"PHL -> CDG"},{"id":3,"label":"PHL -> LHR"}]},{"id":4,"name":"Betsy Ross","tickets":[{"id":5,"label":"PHL -> DCA"}]}]
```

**Comment by vkostyukov** [comment]
Maybe @yanana can quickly spot the error?

**Comment by rpless** [comment]
Yeah, so far it just seems to be strings. I've pushed my current work [here](https://github.com/rpless/FrameworkBenchmarks/blob/update-finch/frameworks/Scala/finch/src/main/scala/WebServer.scala). Might be a good starting put. I'll continue to look at it, but help other sets of eyes are appreciated.

**Comment by travisbrown** [comment]
Argonaut provides a codec for strings that quotes them, and that's probably what's getting picked up here. As a workaround you could provide the instance in `EncodeResponse` explicitly. In the longer term we may want to consider parametrizing `EncodeResponse` on the contentType (at the type level).

**Comment by rpless** [comment]
Ok, so it seems I wasn't importing `io.finch.response.EncodeResponse._`, which means I didn't have the `encodeString` implicit on scope. Importing that package fixes the issue. I find this more than a little concerning because I don't know what implicit conversions are making the other behavior happen. I think we probably need to note this somewhere.

**Comment by travisbrown** [comment]
You shouldn't need to import `EncodeResponse._` in normal usage—the compiler knows to look for the `EncodingResponse[String]` instance there. The issue is that `io.finch.argonaut` also provides an `EncodeResponse[String]`, and that one takes priority (until you import the original one explicitly).

More generally, it makes sense to want to have different `EncodeResponse` instances for the same type depending on the content type. This wouldn't be difficult, but it would be a pretty big API change.

**Comment by yanana** [comment]
Exactly as @travisbrown said, an `EncodeResponse` which converts String to JSON is used as `Content-Type: application/json;charset=utf-8` suggests.

**Comment by rpless** [comment]
I don't think it has anything to do with argonaut. Especially seeing as the plain text endpoint doesn't use it. I just removed both EncodeResponse._ and all references to argonaut in that file and I still get the original behavior.

**Comment by travisbrown** [comment]
Which original behavior do you mean?

Here's an example of how the Argonaut instance can change things:

```scala
scala> import io.finch.response.EncodeResponse
import io.finch.response.EncodeResponse

scala> implicitly[EncodeResponse[String]].apply("foo")
res0: com.twitter.io.Buf = ByteBuffer(3)

scala> import io.finch.argonaut._
import io.finch.argonaut._

scala> implicitly[EncodeResponse[String]].apply("foo")
res1: com.twitter.io.Buf = ByteBuffer(5)
```

The second `Buf` is two characters too long.

**Comment by rpless** [comment]
Disregard that last message, you're right it is argonaut causing the issue.

**Comment by rpless** [comment]
Ok, so then this isn't a bug with code, but more of an issue of me not being aware of all of the EncodeResponses on scope. We probably need to address this. Do you want to close the issue and open a more general one for parameterizing `EncodeResponse` or keep everything here?

**Comment by vkostyukov** [comment]
@travisbrown nice analysis!
@yanana good catch of the `application/json` content-type (I somehow missed), which indicates that a wrong instance of `EncodeResponse` has been used.

For now it's good to know a workaround: an explicit import of the implicit conversion raises its priority. But we definitely have to think about type-level `EncodeResponse`.

**Comment by rpless** [comment]
Thanks for figuring this out @travisbrown and @yanana! I'm going to close this, as this particular issue is fixed.

**Comment by vkostyukov** [comment]
@rpless maybe we should keep this opened until we introduce the new `EncodeResponse` API? Although, I'm totally fine with opening a new issue for that.

**Comment by rpless** [comment]
@vkostyukov My feeling is that this issue is rather specific and the idea of a new `EncodeResponse` API is only introduced about halfway through the thread. A more generic issue might better represent the problem without the noise from this issue (it can certainly reference this issue though). But I'm fine with leaving it here. It's your call.

**Comment by travisbrown** [comment]
I think the way to go on this is a type member representing the content type inside `EncodeResponse`—that way most usage is unchanged, but you can ask for an encoder for a specific content type when you want one.

I'll take a stab at an implementation this evening.
